2.5.0

KillerOfPie opened the pull request:
[contributing]: https://github.com/SparklingComet/TradeShop/blob/master/.github/CONTRIBUTING.md

## Pull Request Etiquette

There are several guidelines you should follow in order for your Pull Request to be merged.
Replace `[ ]` with `[x]` to cross the checkbox.

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

> It is sometimes better to include more changes in a single commit. 
  If you find yourself having an overwhelming amount of commits, you
  can **rebase** your branch.

## Changes

_Replace this sentence with a description of che changes introduced by this PR._

KillerOfPie committed:
Added prevent-trade permission to prevent players from trading with shops
- #117

KillerOfPie committed:
Add Improved Illegal Items from #115
- Initial Add of colour widening for items starting with `@`
- Committing midway so I can copy the basic add of ^ if necessary later
- Updated Config to v. 1.2 and added fix for old ItemBlackLists

KillerOfPie committed:
Finished Implementation of Colour Widening for Illegal Lists
- Fixed some small bugs and poorly ordered calls
- Minor testing done

KillerOfPie committed:
Added Unlimited Admin setting #114
- Setting should allow any player with Admin enabled to use any command to manage any shop

KillerOfPie committed:
Minor Changes
- Internal Blacklisted items are now removed if present when Global Illegal List is set to Whitelist
- Added PreComment NewLine fix to SettingSectionKeys

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enumys/ListType.java:
```diff
@@ -0,0 +1,34 @@
+    WHITELIST,
+    DISABLED
+
+}
```
I think this could be a static nested class (maybe of Shop?), as the low complexity of it doesn't warrant a while new source file

KillerOfPie replied:
See below, if it does get nested it would be in the new IllegalList object

KillerOfPie replied:
Threw this in IllegalItemList, and all seems to work fine if need be I can always move it back later

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enumys/TradeItemType.java:
```diff
@@ -0,0 +1,33 @@
+    COST,
+    PRODUCT
+
+}
```
Same here, this could be a static nested class of ShopItemStack or something similar

KillerOfPie replied:
it's used in multiple classes outside of that, nesting it would limit access to it if I'm not mistaken. Also I really didn't like having the setting and message section enums nested which is why I did it this way

KillerOfPie replied:
Also put this in IllegalItemList for now since that's the only system that uses it for now.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/IllegalItemList.java:
```diff
@@ -0,0 +1,146 @@
+            }
+        }
+    }
+}
```
Has this system been tested?

KillerOfPie replied:
I did basic testing on it and had no problems with any lists/moses

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/Shop.java:
```diff
@@ -36,10 +36,7 @@
-import org.shanerx.tradeshop.enumys.ShopRole;
-import org.shanerx.tradeshop.enumys.ShopStatus;
-import org.shanerx.tradeshop.enumys.ShopType;
+import org.shanerx.tradeshop.enumys.*;
```
Not a fan of this (wildcard imports)

KillerOfPie replied:
Neither am I, my settings in idea may be wonky. But also it's only for our own files so I wasn't too concerned about it.

KillerOfPie replied:
fixed in my idea and re-optimized all the imports so they should all be gone, messaged in discord for the settings that needed to be changed.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/Shop.java:
```diff
@@ -1037,7 +1034,7 @@ else if (shopType == ShopType.BITRADE)
+			if (utils.isIllegal(TradeItemType.COST, iS.getItemStack().getType()))
 				return false;
 		}
 
```
indentation is kinda weird here

KillerOfPie replied:
Fixed in the next commit

SparklingComet reviewed (commented):
Not sure I understand correctly, but it seems to me that if `unlimited-admin` is set to `true`, this means players with higher privileges can do what they want (which is what was requested).

However, I think we should add a step in between, namely that doing so still requires admin mode enabled for that specific player. That way, admins can both serve as staff members and play play the game as regular users without the two things interfering with one another.

KillerOfPie commented:
> Not sure I understand correctly, but it seems to me that if `unlimited-admin` is set to `true`, this means players with higher privileges can do what they want (which is what was requested).
> 
> However, I think we should add a step in between, namely that doing so still requires admin mode enabled for that specific player. That way, admins can both serve as staff members and play play the game as regular users without the two things interfering with one another.

This is already being done anywhere the unlimited check was added
`(Setting.UNLIMITED_ADMIN.getBoolean() && Permissions.isAdminEnabled(pSender)) `

SparklingComet commented:
> This is already being done anywhere the unlimited check was added
> `(Setting.UNLIMITED_ADMIN.getBoolean() && Permissions.isAdminEnabled(pSender)) `

Alright, thanks for the clarification

KillerOfPie committed:
Re-organization, requested reformat, and removal of * imports

KillerOfPie committed:
Fixing Colour widening not working as expected
- Re-organized Debug Modes
- Use debug mode 3 for startup and list creation info
- Changed widening text from starting with @ to start/end `$`/`^` i.e `$dye^`
- Removed old Utils#log method

SparklingComet committed:
Add expiration MVN profile and builddate.yml resource

KillerOfPie committed:
Added Functionality for DEV version expiration
- The below only happens for DEV builds
- Messages start 1 Month after build date at 10 per day
- 2 Months after build date the plugin will only have 24 hours of run time with the below warning.
- There will be 25 messages per day, 2 time remaining messages that are also broadcast to players with Admin permission, and finally plugin disable with notification sent to console and Admin players.

KillerOfPie committed:
Added Custom Setting and Usage Metrics
- Added Metrics Managed to clean up adding of custom metric charts
- Fixed Potential bug in DataStorage#getShopCountInWorld where worlds(like the nether and end) that exist without shops would throw null if checked
- Fixed bad array count in ShopChunk#deserialize
- Added trade count so metrics can be sent on the amount of trades

KillerOfPie committed:
Message and Setting changes
- Changed TOO_MANY_ITEMS to be clearer
- Added Pre/Post Commant New Line fixing to Message
- Changed PreComment New Line Fix in Setting to affect Pre amd Post Comments
- Added Version and #upgrade to Message
- change Setting#fixUp to Setting#upgrade and did some minor optimizing

KillerOfPie committed:
Unifying Message and Setting common methods + Language Support
- Basic testing complete, including minor upgrade testing
- Normal usage of Setting and Message not changed
- Fixed Setting/Message Upgrades not performing properly if version was not present before upgrade
- Fixed old command sending usages
- Added support for built-in language support to be put to use much later
- Added Metrics for language

KillerOfPie committed:
Bug fixes and Minor Changes
- Fixed bug in Data storage that caused chest-linkage files to be wiped on every launch/reload
- Fixed "bug" where both sides of a shops double chest were not added to the chest linkage map
- Removed Unused messages from Message enum and lang file
- Fixed Bug causing a Shops chest to be unlinked if the other half of a double chest was destroyed
- Changed potential bug in Utils#findShopSign, the sign is now found in the same direction order as the sign would find the chest but with each direction flipped instead of the whole list flipped(could grab a different data from shop->chest vs chest->shop)

KillerOfPie committed:
Bug fixes and Json Data cleanup
- Separated JsonConfiguration for better readability
- Properly added Linkage file management
- Added Variable enum to prevent hand type variables(noticed there were 2 different formats in the message file)
- Fixed Config manager not replacing empty values with defaults in Messages and config
- Fixed formatting error in en-us.yml
- Added fix for old messages files replacing `{` and `}` with `%` for variable
- Changed Message variable replacement to process sent/uppercase/lowercase of the string sent
- Fixed bug in `addcost` item stacking that caused the stacked amount to be added without the original item being removed
- Fixed item names with hex codes not being displayed on signs, no color formatting from item names will appear on signs or messages
- Fixed OP players not being able to trade due to having the preventTrade permission
- Fixed Players with the preventTrade permission seeing the message on any block break

KillerOfPie committed:
Fix for Lang not being loaded on Windows machines

KillerOfPie committed:
Fix for config not properly loading after being created.

KillerOfPie committed:
Fix for null shops in getInvolvedStatuses

KillerOfPie committed:
Added Language fallback to default (UNTESTED)
- If the selected language does not have a default value then pull from the default language(en-us)

KillerOfPie committed:
Bump to 2.5.0-Stable
- Updated messages.yml and settings.yml

KillerOfPie committed:
Fix for launch error with unused lang comment items
